### PR TITLE
Onnx wrapper

### PR DESCRIPTION
**Pull request by mike-ferguson**
ONNX Wrapper PR - Changes still incoming, initial commit.

### CODE REVIEW

File: model_tools/activations/onnx.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+    def __call__(self, *args, **kwargs):  # cannot assign __call__ as attribute due to Python convention
+        return self._extractor(*args, **kwargs)
+
+    def get_activations(self, images, layer_names):
+
+        # create directory to store ONNX models
+        import os
+        if not os.path.exists("ONNX Partial Models"):
+            os.makedirs("ONNX Partial Models")
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:47]: this is very ad-hoc, I would:
1. move the definition of this directory onto the file level 
2. allow this to be changed from environment variables

i.e. `ONNX_MODEL_DIRECTORY = os.getenv('ONNX_MODEL_DIRECTORY', '~/.onnx_models')`

File: model_tools/activations/onnx.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+        # check to make sure model is legitimate
+        onnx_model = self._model
+        model_name = self.identifier
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:51]: these re-assignments are unnecessary and imo more confusing than helpful, just use `self._model` and `self.identifier` in-place

File: model_tools/activations/onnx.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        last_layer = output_names[-1]
+
+        # init activations dict for return:
+        new_dict = {}
+
+        # loop through each layer:
+        for layer in layer_names:
+
+            # handle logits case - get last layer activations
+            if layer_names[0] == 'logits':
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:67]: shouldn't this be `if layer == 'logits':`?

File: model_tools/activations/onnx.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+            # process images:
+            torch_images = [torch.from_numpy(image) for image in images]
+            var_images = Variable(torch.stack(torch_images))
+            images_device = var_images.to(self._device)
+
+            # compute ONNX Runtime output prediction
+            ort_inputs = {sess.get_inputs()[0].name: to_numpy(images_device)}
+            ort_outs = sess.run(None, ort_inputs)
+            activations = ort_outs[0]
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:101]: do I understand this right that for _every single layer_, the images are passed through the model up to that layer? Is there not a more efficient way where the images are passed through the model only once and _all_ layer activations captured in this single pass?

File: model_tools/activations/onnx.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        return repr(self._model)
+
+    # def graph(self):
+    #     import networkx as nx
+    #     g = nx.DiGraph()
+    #     for layer in self._model.layers:
+    #         g.add_node(layer.name, object=layer, type=type(layer))
+    #         for outbound_node in layer._outbound_nodes:
+    #             g.add_edge(layer.name, outbound_node.outbound_layer.name)
+    #     return g
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:119]: dead code

File: model_tools/activations/onnx.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+
+# takes any framework from supported list and converted to ONNX
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:122]: -> takes a model in any supported framework and converts it to ONNX

File: model_tools/activations/onnx.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+def to_onnx(batch_size, in_channel, image_size, model, model_name):
+
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:124]: remove empty line

File: model_tools/activations/onnx.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    # generate dummy input
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:125]: why is this necessary?

File: model_tools/activations/onnx.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    x = torch.randn(batch_size, in_channel, image_size, image_size, requires_grad=True)
+    torch_out = model(x)
+
+    # Export the model to onnx
+    torch.onnx.export(model,                        # model being run
+                      x,                            # model input (or a tuple for multiple inputs)
+                      f"{model_name}.onnx",              # where to save the model (can be a file or file-like object)
+                      export_params=True,           # store the trained parameter weights inside the model file
+                      opset_version=10,             # the ONNX version to export the model to
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:134]: is there a rationale behind this? Why not use the default, i.e. don't pass this argument?

File: model_tools/activations/onnx.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                      do_constant_folding=True,     # whether to execute constant folding for optimization
+                      input_names=['input'],        # the model's input names
+                      output_names=['output'],                   # the model's output names
+                      dynamic_axes={'input': {0: 'batch_size'},  # variable length axes
+                                    'output': {0: 'batch_size'}})
+
+    onnx_model = onnx.load( f"{model_name}.onnx")
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:141]: delete space after `( `

File: model_tools/activations/onnx.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    onnx.checker.check_model(onnx_model)
+    ort_session = onnxruntime.InferenceSession( f"{model_name}.onnx")
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:143]: delete space after `(`

File: model_tools/activations/onnx.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+    def to_numpy(tensor):
+        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:146]: redundantly defined with above

File: model_tools/activations/onnx.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+    # compute ONNX Runtime output prediction
+    ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(x)}
+    ort_outs = ort_session.run(None, ort_inputs)
+
+    # compare ONNX Runtime and PyTorch results
+    np.testing.assert_allclose(to_numpy(torch_out), ort_outs[0], rtol=1e-03, atol=1e-05)
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:153]: oh I see -- this, together with the above run, should probably be its own method

File: model_tools/activations/onnx.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+    print("Exported model has been tested with ONNXRuntime, and the result looks good!")
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:155]: use `logging` rather than print

File: model_tools/activations/onnx.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+def get_layers(onnx_model):
+    layers = []
+    for out in enumerate_model_node_outputs(onnx_model):
+        layers.append(out)
+    return layers
+
+
+def get_final_model(framework, batch_size, in_channels, image_size, model, model_name):
+
+    # print(batch_size, in_channels, image_size)
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:168]: dead code

File: model_tools/activations/onnx.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+
+def get_layers(onnx_model):
+    layers = []
+    for out in enumerate_model_node_outputs(onnx_model):
+        layers.append(out)
+    return layers
+
+
+def get_final_model(framework, batch_size, in_channels, image_size, model, model_name):
+
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:167]: delete empty line(s)

File: model_tools/activations/onnx.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+        print("Pytorch to ONNX Conversion successful.")
+        return onnx_model, layers
+
+    # if model is already onnx, return that.
+    elif framework == "onnx":
+        onnx_model = model
+        layers = get_layers(onnx_model)
+        return onnx_model, layers
+
+    # unknown model format. In the future, I hope to add automatic conversion to ONNX for other platforms
```
Comment (mschrimpf) [inline model_tools/activations/onnx.py:184]: that future should be now for this PR to be merged, otherwise we're not winning anything

**Review by mschrimpf — CHANGES_REQUESTED** [review]

**Comment by mschrimpf** [comment]
I looked through it a bit, but stopped after realizing that the current version of this PR only supports pytorch->onnx. Let me know when tf/keras are also supported and tested
